File: 3456-find-the-maximum-length-of-a-good-subsequence-i/find-the-maximum-length-of-a-good-subsequence-i.java
class Solution {
    public int maximumLength(int[] b, int k) {
        int n = b.length;
        int[][] dp = new int[n + 5][k + 5];
        
        for(int i = 0; i <= n-1; i++) {
            Arrays.fill(dp[i], 1);
            
            for(int j = 0; j <= k; j++) {
                for(int i1 = i-1; i1 >= 0; i1--) {
                    if(b[i1] != b[i]) {
                        if(j != 0) {
                            dp[i][j] = Math.max(dp[i][j], 1 + dp[i1][j-1]);
                        }
                    } else {
                        dp[i][j] = Math.max(dp[i][j], 1 + dp[i1][j]);
                    }
                }
            }
        }
        
        int max = 0; 
        for(int i = 0; i <= n-1; i++) {
            int u = dp[i][k];
            max = Math.max(max, u);
        }
        
        return max;
    }
}

File: 3456-find-the-maximum-length-of-a-good-subsequence-i/find-the-maximum-length-of-a-good-subsequence-i.cpp
class Solution {
public:

    int dp[500][26];

    int solve(vector<int>& nums, int i, int k, int &ans){

        if(i == 0)
        return 1;

        if(dp[i][k] != -1)
        return dp[i][k];

        solve(nums, i-1, k, ans);

        int curans = 0;

        for(int j=i-1;j>=0;j--){

            if(nums[j] == nums[i])
            curans = max(curans, solve(nums, j, k, ans));

        }

        if(k > 0){

            for(int j=i-1;j>=0;j--){

                if(nums[j] != nums[i])
                curans = max(curans, solve(nums, j, k-1, ans));

            }

        }

        ans = max(ans, curans+1);

        return dp[i][k] = curans+1;

    }

    int maximumLength(vector<int>& nums, int k) {
        
        int ans = 1;

        memset(dp, -1, sizeof(dp));

        solve(nums, nums.size()-1, k, ans);

        return ans;

    }
};